<launch>

  <arg name="robot_model"                       default=""/>
  <arg name="robot_name"                        default="locobot"/>
  <arg name="show_lidar"                        default="true"/>
  <arg name="external_urdf_loc"                 default=""/>
  <arg name="external_srdf_loc"                 default=""/>
  <arg name="mode_configs"                      default="$(find interbotix_xslocobot_moveit_interface)/config/modes_all.yaml"/>
  <arg name="use_gazebo"                        default="false"/>
  <arg name="use_actual"                        default="true"/>
  <arg name="use_fake"                          default="false"/>
  <arg name="dof"                               default="5"/>
  <arg name="use_camera"                        default="false"/>
  <arg name="use_moveit_rviz"                   default="true"/>
  <arg name="world_name"                        default="$(find interbotix_xslocobot_gazebo)/worlds/xslocobot_gazebo.world"/>
  <arg name="use_cpp_interface"                 default="false"/>
  <arg name="moveit_interface_gui"              default="true"/>
  <arg name="use_python_interface"              default="false"/>

  <arg name="use_lidar"                        default="true"/>
  <arg name="localization"                      default="false"/>
  <arg name="rtabmap_args"                      default=""/>
  <arg name="use_rtabmapviz"                    default="false"/>
  <arg name="rtabmapviz_args"                   default=""/>
  <arg name="database_path"                     default="~/.ros/rtabmap.db"/>
  <arg name="camera_tilt_angle"                 default="0.2618"/>
  <arg name="use_rviz"                          default="true"/>
  <arg name="base_type"                         default="$(optenv INTERBOTIX_XSLOCOBOT_BASE_TYPE create3)"/> <!-- 'kobuki', 'create3' -->
  <arg name="rviz_frame"                        default="map"/>

  <arg if="$(eval base_type == 'kobuki')"  name="move_base_cmd_vel_topic" default="mobile_base/commands/velocity"/>
  <arg if="$(eval base_type == 'create3')" name="move_base_cmd_vel_topic" default="cmd_vel"/>

  <arg if="$(eval base_type == 'kobuki')"  name="odom_topic" default="mobile_base/odom"/>
  <arg if="$(eval base_type == 'create3')" name="odom_topic" default="odom"/>

  <arg     if="$(arg use_lidar)" name="rtabmap_default_args" default="$(arg rtabmap_args)
                                                                      --RGBD/NeighborLinkRefining true
                                                                      --RGBD/ProximityBySpace true
                                                                      --RGBD/ProximityPathMaxNeighbors 10
                                                                      --RGBD/AngularUpdate 0.01
                                                                      --RGBD/LinearUpdate 0.01
                                                                      --RGBD/LocalRadius 5
                                                                      --RGBD/OptimizeFromGraphEnd false
                                                                      --Grid/FromDepth false
                                                                      --Grid/MaxObstacleHeight 0.7
                                                                      --Grid/RayTracing true
                                                                      --Grid/RangeMax 0
                                                                      --Reg/Force3DoF true
                                                                      --Reg/Strategy 1
                                                                      --Mem/STMSize 30
                                                                      --Icp/VoxelSize 0.05
                                                                      --Icp/CorrespondenceRatio 0.4
                                                                      --Icp/MaxCorrespondenceDistance 0.1"/>

  <arg unless="$(arg use_lidar)" name="rtabmap_default_args" default="$(arg rtabmap_args)
                                                                      --RGBD/NeighborLinkRefining true
                                                                      --RGBD/AngularUpdate 0.01
                                                                      --RGBD/LinearUpdate 0.01
                                                                      --RGBD/LocalRadius 5
                                                                      --RGBD/OptimizeFromGraphEnd false
                                                                      --Grid/FromDepth true
                                                                      --Grid/MaxObstacleHeight 0.7
                                                                      --Grid/RayTracing true
                                                                      --Reg/Force3DoF true
                                                                      --Reg/Strategy 0
                                                                      --Mem/STMSize 30"/>

  <include file="$(find interbotix_xslocobot_moveit)/launch/xslocobot_moveit.launch">
    <arg name="robot_model"                       value="$(arg robot_model)"/>
    <arg name="robot_name"                        value="$(arg robot_name)"/>
    <arg name="show_lidar"                        value="$(arg show_lidar)"/>
    <arg name="external_urdf_loc"                 value="$(arg external_urdf_loc)"/>
    <arg name="external_srdf_loc"                 value="$(arg external_srdf_loc)"/>
    <arg name="mode_configs"                      value="$(arg mode_configs)"/>
    <arg name="use_gazebo"                        value="$(arg use_gazebo)"/>
    <arg name="use_actual"                        value="$(arg use_actual)"/>
    <arg name="use_fake"                          value="$(arg use_fake)"/>
    <arg name="dof"                               value="$(arg dof)"/>
    <arg name="use_camera"                        value="$(arg use_camera)"/>
    <arg name="use_moveit_rviz"                   value="$(arg use_moveit_rviz)"/>
    <arg name="world_name"                        value="$(arg world_name)"/>
  </include>

  <group if="$(arg use_cpp_interface)">

    <node
      name="moveit_interface"
      pkg="interbotix_moveit_interface"
      type="moveit_interface"
      respawn="false"
      output="screen"
      ns="$(arg robot_name)"/>

    <node if="$(arg moveit_interface_gui)"
      name="moveit_interface_gui"
      pkg="interbotix_moveit_interface"
      type="moveit_interface_gui"
      output="screen"
      ns="$(arg robot_name)"/>

  </group>



  <group ns="$(arg robot_name)/rtabmap">

    <node
      pkg="nodelet"
      type="nodelet"
      name="rgbd_sync"
      output="screen"
      args="standalone rtabmap_ros/rgbd_sync">
      <param name="approx_sync"                   value="false"/>
      <remap from="rgb/image"                     to="/$(arg robot_name)/camera/color/image_raw"/>
      <remap from="depth/image"                   to="/$(arg robot_name)/camera/aligned_depth_to_color/image_raw"/>
      <remap from="rgb/camera_info"               to="/$(arg robot_name)/camera/color/camera_info"/>
      <remap from="rgbd_image"                    to="rgbd_image"/>
    </node>

    <node
      pkg="nodelet"
      type="nodelet"
      name="points_xyzrgb"
      output="screen"
      args="standalone rtabmap_ros/point_cloud_xyzrgb">
      <param name="decimation"                    value="4"/>
      <param name="voxel_size"                    value="0.05"/>
      <param name="approx_sync"                   value="false"/>
      <remap from="cloud"                         to="depth/color/voxels"/>
    </node>

    <node
      pkg="nodelet"
      type="nodelet"
      name="obstacle_detection"
      output="screen"
      args="standalone rtabmap_ros/obstacles_detection">
      <param name="wait_for_transform"            value="true"/>
      <param name="frame_id"                      value="$(arg robot_name)/base_footprint"/>
      <remap from="cloud"                         to="depth/color/voxels"/>
      <remap from="ground"                        to="depth/color/ground"/>
      <remap from="obstacles"                     to="depth/color/obstacles"/>
    </node>

    <node
      name="rtabmap"
      pkg="rtabmap_ros"
      type="rtabmap"
      output="screen"
      args="$(arg rtabmap_default_args)">
      <param name="subscribe_depth"               value="false"/>
      <param name="subscribe_rgb"                 value="false"/>
      <param name="subscribe_rgbd"                value="true"/>
      <param name="subscribe_stereo"              value="false"/>
      <param name="subscribe_scan"                value="$(arg use_lidar)"/>
      <param name="subscribe_scan_cloud"          value="false"/>
      <param name="subscribe_scan_descriptor"     value="false"/>
      <param name="subscribe_user_data"           value="false"/>
      <param name="subscribe_odom_info"           value="false"/>
      <param name="frame_id"                      value="$(arg robot_name)/base_footprint"/>
      <param name="map_frame_id"                  value="map"/>
      <param name="odom_frame_id"                 value="$(arg robot_name)/odom"/>
      <param name="publish_tf"                    value="true"/>
      <param name="odom_tf_angular_variance"      value="0.05"/>
      <param name="odom_tf_linear_variance"       value="0.1"/>
      <param name="odom_sensor_sync"              value="false"/>
      <param name="wait_for_transform_duration"   value="0.2"/>
      <param name="database_path"                 value="$(arg database_path)"/>
      <param name="approx_sync"                   value="true"/>
      <param name="queue_size"                    value="10"/>
      <param name="tag_linear_variance"           value="0.0001"/>
      <param name="tag_angular_variance"          value="9999"/>
      <param name="Mem/InitWMWithAllNodes"        value="$(arg localization)"/>
      <param     if="$(arg localization)" name="Mem/IncrementalMemory" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" value="true"/>
      <remap from="scan"                          to="/$(arg robot_name)/scan"/>
      <remap from="initialpose"                   to="/$(arg robot_name)/initialpose"/>
      <remap from="goal_out"                      to="/$(arg robot_name)/move_base_simple/goal"/>
    </node>

    <node if="$(arg use_rtabmapviz)"
      pkg="rtabmap_ros"
      type="rtabmapviz"
      name="rtabmapviz"
      args="$(arg rtabmapviz_args)"
      output="screen">
      <param name="subscribe_rgbd"                value="true"/>
      <param name="subscribe_scan"                value="$(arg use_lidar)"/>
      <param name="frame_id"                      value="$(arg robot_name)/base_footprint"/>
      <param name="odom_frame_id"                 value="$(arg robot_name)/odom"/>
      <param name="wait_for_transform"            value="true"/>
      <remap from="scan"                          to="/$(arg robot_name)/scan"/>
    </node>
  </group>

  <node
    pkg="move_base"
    type="move_base"
    respawn="false"
    name="move_base"
    output="screen"
    ns="$(arg robot_name)">
    <rosparam     if="$(arg use_lidar)" file="$(find interbotix_xslocobot_nav)/config/local_costmap_all_params.yaml" command="load"/>
    <rosparam     if="$(arg use_lidar)" file="$(find interbotix_xslocobot_nav)/config/global_costmap_all_params.yaml" command="load"/>
    <rosparam unless="$(arg use_lidar)" file="$(find interbotix_xslocobot_nav)/config/local_costmap_depth_params.yaml" command="load"/>
    <rosparam unless="$(arg use_lidar)" file="$(find interbotix_xslocobot_nav)/config/global_costmap_depth_params.yaml" command="load"/>
    <rosparam file="$(find interbotix_xslocobot_nav)/config/common_costmap_params.yaml" command="load" ns="global_costmap"/>
    <rosparam file="$(find interbotix_xslocobot_nav)/config/common_costmap_params.yaml" command="load" ns="local_costmap"/>
    <rosparam file="$(find interbotix_xslocobot_nav)/config/local_planner_params.yaml"  command="load"/>
    <rosparam file="$(find interbotix_xslocobot_nav)/config/global_planner_params.yaml" command="load"/>
    <rosparam file="$(find interbotix_xslocobot_nav)/config/move_base_params.yaml"      command="load"/>

    <param name="global_costmap/robot_base_frame" value="$(arg robot_name)/base_footprint"/>
    <param name="local_costmap/global_frame"      value="$(arg robot_name)/odom"/>
    <param name="local_costmap/robot_base_frame"  value="$(arg robot_name)/base_footprint"/>

    <remap from="cmd_vel"                         to="$(arg move_base_cmd_vel_topic)"/>
    <remap from="map"                             to="rtabmap/grid_map"/>
    <remap from="odom"                            to="$(arg odom_topic)"/>
  </node>




  <node if="$(arg use_python_interface)"
    name="moveit_python_interface"
    pkg="interbotix_xslocobot_moveit_interface"
    type="moveit_python_interface"
    respawn="false"
    output="screen"
    ns="$(arg robot_name)">
    <rosparam command="load"                      file="$(find interbotix_xslocobot_moveit_interface)/config/$(arg robot_model).yaml"/>
    <param name="robot_model"                     value="$(arg robot_model)"/>
    <remap from="/attached_collision_object"      to="/$(arg robot_name)/attached_collision_object"/>
    <remap from="/collision_object"               to="/$(arg robot_name)/collision_object"/>
  </node>

</launch>
